File: leetcode/cpp/[4]_median-of-two-sorted-arrays.cpp
/*
 * @lc app=leetcode id=4 lang=cpp
 *
 * [4] Median of Two Sorted Arrays
 */

// @lc code=start
class Solution {
public:
  /*
  * Accepted
  * 2094/2094 cases passed (20 ms)
  * Your runtime beats 93.61 % of cpp submissions
  * Your memory usage beats 73.07 % of cpp submissions (89 MB)
  * 
  * Time complexity : O((m+n)/2)
  * Space complexity : O(1)
  */
  double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
    int num1Pos = 0, num2Pos = 0;
    int prevVal = 0, currVal = 0;

    for (int i = 0; i < ((nums1.size() + nums2.size()) / 2) + 1; i++) {
      if (num2Pos == nums2.size() || (num1Pos < nums1.size() && nums1[num1Pos] < nums2[num2Pos])) {
        prevVal = currVal;
        currVal = nums1[num1Pos++];
      } else {
        prevVal = currVal;
        currVal = nums2[num2Pos++];
      }
    }

    return (nums1.size() + nums2.size()) % 2 == 0 ?(double)(prevVal + currVal) / 2 : currVal;
  }
};
// @lc code=end

File: leetcode/python3/[4]_median-of-two-sorted-arrays.py
#
# @lc app=leetcode id=4 lang=python3
#
# [4] Median of Two Sorted Arrays
#

# @lc code=start
class Solution:
  '''
  Accepted
  2094/2094 cases passed (96 ms)
  Your runtime beats 47.06 % of python3 submissions
  Your memory usage beats 54.94 % of python3 submissions (14.5 MB)

  Time complexity : O((m+n)/2)
  Space complexity : O(1)
  '''
  def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
    nums1Pos, nums2Pos = 0, 0
    prevVal, currVal = 0, 0

    for i in range(int((len(nums1) + len(nums2)) / 2) + 1):
      if nums2Pos == len(nums2) or (nums1Pos < len(nums1) and nums1[nums1Pos] <= nums2[nums2Pos]):
        prevVal = currVal
        currVal = nums1[nums1Pos]
        nums1Pos += 1
      else:
        prevVal = currVal
        currVal = nums2[nums2Pos]
        nums2Pos += 1
    
    return (prevVal + currVal) / 2 if (len(nums1) + len(nums2)) % 2 == 0 else currVal
# @lc code=end

